### PR TITLE
Realtime scheduling

### LINKED ISSUE #964: Support optional real-time scheduling mode
In our new process-based architecture, we may get a performance boost if we can reduce unnecessary context switches by utilizing a non-default CPU scheduler to run the shadow and plugin threads. Using a real-time scheduler like `SCHED_FIFO` will give us tighter control over when context switches will occur.

We think it's worth supporting a real-time scheduling mode as an optional run-mode, so that we can test if it provides a performance boost. In order for us to be allowed to use a real-time mode, support needs to be configured on the simulation host via ulimit.

More info:
- https://man7.org/linux/man-pages/man7/sched.7.html
- https://man7.org/linux/man-pages/man2/sched_setscheduler.2.html

### PR DESCRIPTION
**Pull request by sporksmith**
* Adds some sched_yields inside shadow to prevent poor performance or deadlocks under CPU contention when using the realtime scheduler. These also make performance less-bad when there's CPU contention, even with the regular scheduler.
* Adds a command-line option for using the realtime scheduler

Some initial analysis at https://github.com/sporksmith/phantom-paper/blob/realtime-experiment/experiments/realtime/analysis.ipynb

More analysis in progress; looks like cpu-pinning+realtime is a significant win in preload mode.

The ptrace modes are currently broken with the realtime scheduler.

Fixes #964

### CODE REVIEW

File: src/main/core/main.c [hunk 1]
```diff
@@ -250,6 +259,19 @@ gint main_runShadow(gint argc, gchar* argv[]) {
         }
     }
 
+    if (_setSchedFifo) {
+        struct sched_param param = {0};
+        param.sched_priority = 1;
+        int rc = sched_setscheduler(0, SCHED_FIFO, &param);
+
+        if (rc != 0) {
+            fprintf(stderr, "Could not set SCHED_FIFO!\n");
```
Comment (robgjansen) [inline src/main/core/main.c:268]: The logger has already been set up at this point, so can we log this with our logger? How about:
```
error("Could not set real-time scheduler mode to SCHED_FIFO!")
```

**Review by robgjansen — APPROVED** [review]
Easy-peasy. I just have some minor suggested changes. Thanks!

File: src/main/core/main.c [hunk 2]
```diff
@@ -26,6 +27,14 @@
 #include "support/logger/logger.h"
 #include "main/bindings/c/bindings.h"
 
+// We use an int here because the option parsing library doesn't provide a way
+// to set a boolean flag to false explicitly.
+static gint _setSchedFifo = false;
+OPTION_EXPERIMENTAL_ENTRY(
+    "set-sched-fifo", 0, 0, G_OPTION_ARG_INT, &_setSchedFifo,
+    "Use the SCHED_FIFO scheduler. Requires CAP_SYS_NICE. See sched(7), capabilities(7) [0]",
+    "[0|1]")
```
[the remark this answers is not in the document]
Comment (sporksmith) [reply src/main/core/main.c]: Ah I see. I'd previously looked for but couldn't find a way to suppoert e.g. `--set-sched-fifo=true` and `--set-sched-fifo=false`. It looks like if we wanted to add a way to explicitly disable it we could add another option `--no-set-sched-fifo` using the option `G_OPTION_FLAG_REVERSE`

File: src/main/core/main.c (hunk 1 above)
Comment (sporksmith) [reply src/main/core/main.c]: Done

File: src/main/core/main.c [hunk 3]
```diff
@@ -250,6 +259,19 @@ gint main_runShadow(gint argc, gchar* argv[]) {
...
+    if (_setSchedFifo) {
+        struct sched_param param = {0};
+        param.sched_priority = 1;
+        int rc = sched_setscheduler(0, SCHED_FIFO, &param);
+
+        if (rc != 0) {
+            fprintf(stderr, "Could not set SCHED_FIFO!\n");
+            return -1;
+        } else {
+            fprintf(stderr, "In mode SCHED_FIFO.\n");
```
[the remark this answers is not in the document]
Comment (sporksmith) [reply src/main/core/main.c]: Done

**Pull request merged**